### PR TITLE
Add flexmock, sqlalchemy-utils, pgpy, and py-gfm

### PR DESCRIPTION
**Pull request by bsima**
###### Motivation for this change

Trying to build [core.sr.ht](https://git.sr.ht/~sircmpwn/core.sr.ht) for NixOS and running into missing packages, figured I would add them upstream. I tested each one with `nix-build -A pythonPackages.<pkgname>` to make sure they built.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by lsix — CHANGES_REQUESTED** [review]
Hi,

Thanks for the PR.

Could you add `meta` attributes for `flexmock`, `PGPy` and `py-gfm` ?

**Comment by bsima** [comment]
Lancelot SIX <notifications@github.com> writes:

> lsix requested changes on this pull request.
>
> Hi,
>
> Thanks for the PR.
>
> Could you add `meta` attributes for `flexmock`, `PGPy` and `py-gfm` ?

Okay, I added the `meta` attributes, and added one more package,
'sshpubkeys'. Let me know if you need any more changes :)

File: pkgs/development/python-modules/py-gfm/default.nix [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "ef6750c579d26651cfd23968258b604228fd71b2a4e1f71dea3bea289e01377e";
+  };
+
+  buildInputs = [ markdown ];
+
+  meta = with lib; {
+    homepage = https://github.com/zopieux/py-gfm;
+    description = "This is an implementation of GitHub-Flavored Markdown written as an extension to the Python Markdown library. It aims for maximal compatibility with GitHub's rendering.";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/py-gfm/default.nix]: This description is too long.

**Comment by bsima** [comment]
worldofpeace <notifications@github.com> writes:

> +  meta = with lib; {
> +    homepage = https://github.com/zopieux/py-gfm;
> +    description = "This is an implementation of GitHub-Flavored Markdown written as an extension to the Python Markdown library. It aims for maximal compatibility with GitHub's rendering.";
>
> This description is too long.

Fixed in 9e0430cb919

File: pkgs/development/python-modules/flexmock/default.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+buildPythonPackage rec {
+  pname = "flexmock";
+  version = "0.10.2";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "fe95c8727f4db73dc8f2f7b4548bffe7992440a965fefd60da291abda5352c2b";
+  };
+
+  buildInputs = [ nose pytest ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/flexmock/default.nix]:
```suggestion
  checkInputs = [ nose pytest ];
```

File: pkgs/development/python-modules/flexmock/default.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
+    license = licenses.bsd2;
+  };
```
Comment (dotlambda) [inline pkgs/development/python-modules/flexmock/default.nix]: `maintainers`?

File: pkgs/development/python-modules/pgpy/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+buildPythonPackage rec {
+  pname = "PGPy";
+  version = "0.4.3";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "04412dddd6882ac0c5d5daf4326c28d481421851a68e25e7ac8e06cc9dc2b902";
+  };
+
+  buildInputs = [ cryptography pyasn1 six singledispatch ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/pgpy/default.nix]:
```suggestion
  propagatedBuildInputs = [ cryptography pyasn1 six singledispatch ];
```

File: pkgs/development/python-modules/pgpy/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+  meta = with lib; {
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "PGPy is a Python (2 and 3) library for implementing Pretty Good Privacy into Python programs, conforming to the OpenPGP specification per RFC 4880.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pgpy/default.nix]: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

File: pkgs/development/python-modules/pgpy/default.nix [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    license = licenses.bsd3;
+  };
```
Comment (dotlambda) [inline pkgs/development/python-modules/pgpy/default.nix]: `maintainers`?

File: pkgs/development/python-modules/flexmock/default.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "fe95c8727f4db73dc8f2f7b4548bffe7992440a965fefd60da291abda5352c2b";
+  };
+
+  buildInputs = [ nose pytest ];
+
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/flexmock/default.nix]: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

**Review by dotlambda — CHANGES_REQUESTED** [review]
The requested changes should be done for the packages I didn't comment on as well.
You might want to read through https://nixos.org/nixpkgs/manual/#python again.

**Comment by bsima** [comment]
Robert Schütz <notifications@github.com> writes:

> dotlambda requested changes on this pull request.
>
> The requested changes should be done for the packages I didn't comment on as well.

Okay, I think I covered all of them. They all have meta, metaintainers, etc.

> You might want to read through https://nixos.org/nixpkgs/manual/#python again.

Looking at the contributing guidelines, this is what I see:

> Python libraries are called from python-packages.nix and packaged with
> buildPythonPackage. The expression of a library should be in
> pkgs/development/python-modules/<name>/default.nix. Libraries in
> pkgs/top-level/python-packages.nix are sorted quasi-alphabetically to
> avoid merge conflicts.

Check.

> Python applications live outside of python-packages.nix and are packaged
> with buildPythonApplication.

None of these are apps.

> Make sure libraries build for all Python interpreters.

I think they all succeed. I'll look into testing this thoroughly over
the weekend or next week.

> By default we enable tests. Make sure the tests are found and, in the
> case of libraries, are passing for all interpreters. If certain tests
> fail they can be disabled individually. Try to avoid disabling the tests
> altogether. In any case, when you disable tests, leave a comment
> explaining why.

Only one package failed to find the tests, and I added a comment
explaining that.

> Commit names of Python libraries should reflect that they are Python
> libraries, so write for example pythonPackages.numpy: 1.11 -> 1.12.

Check.

> Attribute names in python-packages.nix should be normalized according to
> PEP 0503. This means that characters should be converted to lowercase
> and . and _ should be replaced by a single - (foo-bar-baz instead of
> Foo__Bar.baz )

Check.

Okay, anything else I can do?

**Comment by bsima** [comment]
@dotlambda Anything I can do to help move this along?

**Review by worldofpeace — CHANGES_REQUESTED** [review]
You don`t need to set any of these meta attributes

```
downloadPage = foo;
priority = 0;
broken = false;
updateWalker = true;
```

`downloadPage` could be useful but we already know where we can download python packages from (pypi.org).

Additionally you've set `platforms` to `all`. This is already set to automatically to https://github.com/NixOS/nixpkgs/blob/1b6a4d3979f13f990617d32ee9ec99b1f02673ec/pkgs/development/interpreters/python/mk-python-derivation.nix#L106
which should be sufficient.

**Comment by worldofpeace** [comment]
@bsima 
Could you also `fixup` these additional commits with an interactive rebase?
It would make reviewing commit by commit possible.

**Comment by bsima** [comment]
worldofpeace <notifications@github.com> writes:

> @bsima 
> Could you also `fixup` these additional commits with an interactive rebase?
> It would make reviewing commit by commit possible.

Yes, I rebased to make things more clear, and updated the meta
info. Thanks for the feedback, and let me know if there is anything else
I can do.

File: pkgs/development/python-modules/flexmock/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "flexmock";
+  version = "0.10.2";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "fe95c8727f4db73dc8f2f7b4548bffe7992440a965fefd60da291abda5352c2b";
+  };
+
+  checkInputs = [ nose pytest ];
+
```
Comment (worldofpeace) [inline pkgs/development/python-modules/flexmock/default.nix]: Noticed the tests weren't being ran.

This should fix that:
```suggestion
  checkPhase = ''
    pytest tests/
  '';
```

File: pkgs/development/python-modules/flexmock/default.nix [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/flexmock/default.nix]: `description` should not have the name of said software in it or a period at the end.
> https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes
> Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does. 

```suggestion
    description = "A testing library for Python that makes it easy to create mocks, stubs, and fakes";
```

File: pkgs/development/python-modules/flexmock/default.nix [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    longDescription = ''
```
Comment (worldofpeace) [inline pkgs/development/python-modules/flexmock/default.nix]: This description isn't long so we don't need it.

File: pkgs/development/python-modules/pgpy/default.nix [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+, fetchPypi
+, buildPythonPackage
+, cryptography
+, pyasn1
+, six
+, singledispatch
+}:
+
+buildPythonPackage rec {
+  pname = "PGPy";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/pgpy/default.nix]: I think we're normalizing `pname` now so this should be
```suggestion
  pname = "pgpy";
```

File: pkgs/development/python-modules/pgpy/default.nix [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  version = "0.4.3";
+
+  src = fetchPypi {
+    inherit pname version;
```
Comment (worldofpeace) [inline pkgs/development/python-modules/pgpy/default.nix]: Changing the `pname` will most likely make this not work here.
So we might need:
```suggestion
    pname = "PGPy";
    inherit version;
```

File: pkgs/development/python-modules/pgpy/default.nix [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    sha256 = "04412dddd6882ac0c5d5daf4326c28d481421851a68e25e7ac8e06cc9dc2b902";
+  };
+
+  propogatedBuildInputs = [ cryptography pyasn1 six singledispatch ];
+
+  meta = with lib; {
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "Pretty Good Privacy for Python 2 and 3.";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/pgpy/default.nix]:
```suggestion
    description = "Pretty Good Privacy for Python 2 and 3";
```

File: pkgs/development/python-modules/pgpy/default.nix [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+buildPythonPackage rec {
+  pname = "PGPy";
+  version = "0.4.3";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "04412dddd6882ac0c5d5daf4326c28d481421851a68e25e7ac8e06cc9dc2b902";
+  };
+
+  propogatedBuildInputs = [ cryptography pyasn1 six singledispatch ];
```
Comment (worldofpeace) [inline pkgs/development/python-modules/pgpy/default.nix]: See https://github.com/SecurityInnovation/PGPy/blob/f1c3d68e32c334f5aa14c34580925e97f17f4fde/setup.py#L36
> only depend on enum34 if Python is older than 3.4

We can constrain the addition of that with
```suggestion
  propogatedBuildInputs = [ cryptography pyasn1 six singledispatch ] ++ lib.optional (pythonOlder "3.4") enum34;
```

File: pkgs/development/python-modules/py-gfm/default.nix [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+buildPythonPackage rec {
+  pname = "py-gfm";
+  version = "0.1.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "ef6750c579d26651cfd23968258b604228fd71b2a4e1f71dea3bea289e01377e";
+  };
+
+  buildInputs = [ markdown ];
```
Comment (worldofpeace) [inline pkgs/development/python-modules/py-gfm/default.nix]:
```suggestion
  propagatedBuildInputs = [ markdown ];
```

File: pkgs/development/python-modules/sqlalchemy-utils/default.nix [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+, sqlalchemy
+, pytest
+, flexmock
+, mock
+, python-dateutil
+, pytz
+}:
+
+buildPythonPackage rec {
+  pname = "SQLAlchemy-Utils";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sqlalchemy-utils/default.nix]: I think we're normalizing pname now so this should be
```suggestion
  pname = "sqlalchemy-utils";
```

File: pkgs/development/python-modules/sqlalchemy-utils/default.nix [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  version = "0.33.6";
+
+  src = fetchPypi {
+    inherit pname version;
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sqlalchemy-utils/default.nix]: Changing the pname will most likely make this not work here.
So we might need:
```suggestion
    pname = "SQLAlchemy-Utils";
    inherit version;
```

File: pkgs/development/python-modules/sqlalchemy-utils/default.nix [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    sha256 = "45ab41c90bfb8dd676e83179be3088b3f2d64b613e3b590187163dd941c22d4c";
+  };
+  checkInputs = [ pytest flexmock mock python-dateutil pytz ];
+  buildInputs = [ six sqlalchemy ];
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sqlalchemy-utils/default.nix]:
```suggestion
  propagatedBuildInputs = [ six sqlalchemy ];
```

File: pkgs/development/python-modules/sqlalchemy-utils/default.nix [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+  meta = with lib; {
+    homepage = https://github.com/kvesteri/sqlalchemy-utils;
+    description = "Various utility functions and datatypes for SQLAlchemy.";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sqlalchemy-utils/default.nix]:
```suggestion
    description = "Various utility functions and datatypes for SQLAlchemy";
```

File: pkgs/development/python-modules/sshpubkeys/default.nix [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+buildPythonPackage rec {
+  pname = "sshpubkeys";
+  version = "3.1.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "b388399caeeccdc145f06fd0d2665eeecc545385c60b55c282a15a022215af80";
+  };
+
+  buildInputs = [ cryptography ecdsa ];
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sshpubkeys/default.nix]:
```suggestion
  propagatedBuildInputs = [ cryptography ecdsa ];
```

File: pkgs/development/python-modules/sshpubkeys/default.nix [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+  # For some reason nix can't find the "tests" module.
+  doCheck = false;
```
Comment (worldofpeace) [inline pkgs/development/python-modules/sshpubkeys/default.nix]: Seems somethings not right with the pypi tarball.
Everything works fine when I use the github repo.

```nix
src = fetchFromGitHub {
    owner = "ojarva";
    repo = "python-sshpubkeys";
    rev = "v${version}";
    sha256 = "0000000000000000000000000000000000000000000000000000"; # change this
};
```

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Additionally tests need aren't running or are missing `checkInputs` for:
 - py-gfm
- pgpy

File: pkgs/development/python-modules/sshpubkeys/default.nix (hunk 21 above)
Comment (worldofpeace) [reply pkgs/development/python-modules/sshpubkeys/default.nix]: Please also add a comment above it explaining why we have to use github.

**Comment by bsima** [comment]
Closing in lieu of #54425 - much more progress and better code in that one :)

**Pull request closed without merging**